crates/biome_string_case/src/lib.rs

```diff
@@ -447,48 +447,81 @@ pub trait Collator {
         let mut iter2 = iter2.into_iter();
         loop {
             match (iter1.next(), iter2.next()) {
-                (Some(c1), Some(c2)) if c1 == c2 => {}
+                (None, None) => {
+                    // All characters of `iter1` and `iter2` are equal.
+                    return Ordering::Equal;
+                }
+                (None, Some(_)) => {
+                    // `iter1` is a prefix of `iter2`.
+                    return Ordering::Less;
+                }
+                (Some(_), None) => {
+                    // `iter2` is a prefix of `iter1`.
+                    return Ordering::Greater;
+                }
+                (Some(c1), Some(c2)) if c1 == c2 => {
+                    // For now, all characters of `iter1` and `iter2` are equal.
+                }
                 (Some(mut c1), Some(mut c2)) => {
-                    if let (Some(n1), Some(n2)) =
-                        (self.as_ascii_digit(&c1), self.as_ascii_digit(&c2))
-                    {
-                        // Compare numbers
-                        // We don't skip leading zeroes.
-                        let mut number_ordering = n1.cmp(&n2);
-                        loop {
-                            match (iter1.next(), iter2.next()) {
-                                (None, None) => {
-                                    return number_ordering;
-                                }
-                                (None, Some(_)) => {
-                                    return Ordering::Less;
-                                }
-                                (Some(_), None) => {
-                                    return Ordering::Greater;
-                                }
-                                (Some(next1), Some(next2)) => {
-                                    c1 = next1;
-                                    c2 = next2;
+                    let mut number_ordering = Ordering::Equal;
+                    // Compare numbers
+                    // We don't skip leading zeroes.
+                    loop {
+                        match (self.as_ascii_digit(&c1), self.as_ascii_digit(&c2)) {
+                            (None, None) => {
+                                if number_ordering == Ordering::Equal {
+                                    // The numbers are equal, we have to compare `c1` and `c2` that are not digits.
+                                    break;
                                 }
+                                return number_ordering;
                             }
-                            match (self.as_ascii_digit(&c1), self.as_ascii_digit(&c2)) {
-                                (Some(n1), Some(_n2)) => {
```

Comment:
A part of the bug was coming from here: note the underscore `_` in front of `n2`.